### PR TITLE
feat!: replace registration tokens by JIT config for ephemeral runners

### PR DESCRIPTION
**Pull request by GuptaNavdeep1983**
# Problem

This PR addresses the issue identified as per the following security advisory:

https://github.com/philips-labs/terraform-aws-github-runner/security/advisories/GHSA-c7m9-5vcx-35m6

# Solution:

GitHub has released a couple of new API's:
 - API for Org level registration:
https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-configuration-for-a-just-in-time-runner-for-an-organization
 - API for Repo level registration:
https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-configuration-for-a-just-in-time-runner-for-a-repository

which can be used to generate encoded just-in-time(JIT) config. This config can then be used to register-and-start the runner. Since this config can be used only once, hence the vulnerability associated with the exploitation of the Github Registration token is automatically resolved.

# Limitations:

This solution is only targeted towards `Ephemeral Runners` which can use this new API. 
The configuration option to start the runner in `service (Non-Ephemeral) runners` will continue to be vulnerable and its `not addressed` in this PR.

# Migrations

When using ephemeral runners with an own start script, you have to [update the start script](https://github.com/philips-labs/terraform-aws-github-runner/pull/3350/files#diff-406ac97629bbe42f24c5f024e74a04e7c217adfd3890a7d5b4aeb6491edfcb3f) to use the JIT config when starting the runner. 

1. Remove the line with `./config.sh` from the script
2. Update the line to start the runner to use the JIT config, `./run.sh --jitconfig $${config}`

# Improvements:

- [ ] https://github.com/philips-labs/terraform-aws-github-runner/issues/3379

# Links:

Please find more relevant links below:
Github changelog link associated with this api: https://github.blog/changelog/2023-06-02-github-actions-just-in-time-self-hosted-runners/
Github documentation link: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-just-in-time-runners

### CODE REVIEW

File: examples/multi-runner/main.tf [hunk 1]
```diff
@@ -17,31 +17,8 @@ module "base" {
 }
 
 module "multi-runner" {
-  source              = "../../modules/multi-runner"
-  multi_runner_config = local.multi_runner_config
-  #  Alternative to loading runner configuration from Yaml files is using static configuration:
```
Comment (GuptaNavdeep1983) [inline examples/multi-runner/main.tf:22]: removed commented code.

File: lambdas/functions/control-plane/src/aws/runners.d.ts [hunk 2]
```diff
@@ -26,11 +26,9 @@ export interface ListRunnerFilters {
 }
 
 export interface RunnerInputParameters {
```
Comment (GuptaNavdeep1983) [inline lambdas/functions/control-plane/src/aws/runners.d.ts:28]: runners module doesnt create the ssm parameter anymore, hence removed those fields.

File: lambdas/functions/control-plane/src/aws/runners.ts [hunk 3]
```diff
@@ -8,7 +8,7 @@ import {
   TerminateInstancesCommand,
```
Comment (GuptaNavdeep1983) [inline lambdas/functions/control-plane/src/aws/runners.ts:8]: creation of ssm parameters have moved to scale-up module.

File: lambdas/functions/control-plane/src/aws/runners.ts [hunk 4]
```diff
@@ -233,22 +220,7 @@ export async function createRunner(runnerParameters: Runners.RunnerInputParamete
...
-      runnerParameters.runnerServiceConfig.join(' '),
-      true,
-    );
-
-    if (isDelay) {
-      // Delay to prevent AWS ssm rate limits by being within the max throughput limit
-      await delay(25);
-    }
-  }
+  return instances;
```
Comment (GuptaNavdeep1983) [inline lambdas/functions/control-plane/src/aws/runners.ts:223]: moved to scale-up module.

File: lambdas/functions/control-plane/src/scale-runners/scale-up.ts [hunk 5]
```diff
@@ -116,21 +138,63 @@ async function isJobQueued(githubInstallationClient: Octokit, payload: ActionReq
   return isQueued;
 }
 
+async function getRunnerGroupId(githubRunnerConfig: CreateGitHubRunnerConfig, ghClient: Octokit): Promise<number> {
```
Comment (GuptaNavdeep1983) [inline lambdas/functions/control-plane/src/scale-runners/scale-up.ts:141]: this function gets the runner group id for the given runner group name either from github or from ssm, attempts to get it from ssm first and if not found from github.

File: lambdas/functions/control-plane/src/scale-runners/scale-up.ts [hunk 6]
```diff
@@ -116,21 +138,63 @@ async function isJobQueued(githubInstallationClient: Octokit, payload: ActionReq
...
+        false,
+      );
+    } else {
+      runnerGroupId = parseInt(runnerGroup);
+    }
+  }
+  return runnerGroupId;
+}
+
+async function GetRunnerGroupByName(ghClient: Octokit, githubRunnerConfig: CreateGitHubRunnerConfig): Promise<number> {
```
Comment (GuptaNavdeep1983) [inline lambdas/functions/control-plane/src/scale-runners/scale-up.ts:171]: gets the runner group name from github.

File: lambdas/functions/control-plane/src/scale-runners/scale-up.ts [hunk 7]
```diff
@@ -116,21 +138,63 @@ async function isJobQueued(githubInstallationClient: Octokit, payload: ActionReq
...
+
+  return runnerGroupId;
+}
+
 export async function createRunners(
   githubRunnerConfig: CreateGitHubRunnerConfig,
   ec2RunnerConfig: CreateEC2RunnerConfig,
   ghClient: Octokit,
 ): Promise<void> {
-  const token = await getGithubRunnerRegistrationToken(githubRunnerConfig, ghClient);
```
Comment (GuptaNavdeep1983) [inline lambdas/functions/control-plane/src/scale-runners/scale-up.ts:124]: moved from here to the function responsible for creating token.

File: lambdas/functions/control-plane/src/scale-runners/scale-up.ts [hunk 8]
```diff
@@ -116,21 +138,72 @@ async function isJobQueued(githubInstallationClient: Octokit, payload: ActionReq
...
 
+async function getRunnerGroupId(githubRunnerConfig: CreateGitHubRunnerConfig, ghClient: Octokit): Promise<number> {
+  // if the runnerType is Repo, then runnerGroupId is default to 1
+  let runnerGroupId: number | undefined = 1;
+  if (githubRunnerConfig.runnerType === 'Org' && githubRunnerConfig.runnerGroup !== undefined) {
+    let runnerGroup: string | undefined;
+    // check if runner group id is already stored in SSM Parameter Store and
+    // use it if it exists to avoid API call to GitHub
+    try {
+      runnerGroup = await getParameter(
```
Comment (npalm) [inline lambdas/functions/control-plane/src/scale-runners/scale-up.ts:149]: This still requires an extra call to SSM for every runner we create. An idea for improment ould be whne the groupId is provided we not store it in ssm. What do you think, address this in an issue as improvement?

File: lambdas/functions/control-plane/src/scale-runners/scale-up.ts [hunk 9]
```diff
@@ -232,3 +308,91 @@ export async function scaleUp(eventSource: string, payload: ActionRequestMessage
...
+        : await ghClient.actions.generateRunnerJitconfigForRepo({
+            owner: githubRunnerConfig.runnerOwner.split('/')[0],
+            repo: githubRunnerConfig.runnerOwner.split('/')[1],
+            name: ephemeralRunnerConfig.runnerName,
+            runner_group_id: ephemeralRunnerConfig.runnerGroupId,
+            labels: ephemeralRunnerConfig.runnerLabels,
+          });
+    // store jit config in ssm parameter store
+    logger.debug('Runner JIT config for ephemeral runner', {
+      runner_jit_config: runnerConfig.data.encoded_jit_config,
```
Comment (npalm) [inline lambdas/functions/control-plane/src/scale-runners/scale-up.ts:389]: Is the jit config sensible data?

File: modules/runners/variables.tf [hunk 10]
```diff
@@ -214,8 +214,8 @@ variable "runner_boot_time_in_minutes" {
   default     = 5
 }
 
-variable "runner_extra_labels" {
-  description = "Extra labels for the runners (GitHub). Separate each label by a comma"
+variable "runner_labels" {
+  description = "All the labels for the runners (GitHub) including the default one's(e.g: self-hosted, linux, x64, label1, label2). Separate each label by a comma"
   type        = string
   default     = ""
```
Comment (npalm) [inline modules/runners/variables.tf:220]: Should we not make the variable required now we expect ALL labels?

File: lambdas/functions/control-plane/src/aws/runners.test.ts [hunk 11]
```diff
@@ -205,91 +200,13 @@ describe('create runner', () => {
...
-    for (const instance of instances[0].InstanceIds) {
-      expect(mockSSMClient).toHaveReceivedCommandWith(PutParameterCommand, {
-        Name: `${SSM_TOKEN_PATH}/${instance}`,
-        Type: 'SecureString',
-        Value: '--token foo --url http://github.com',
-      });
-    }
-  });
-
-  it('calls create fleet of 40 instances (ssm rate limit condition) to test time delay ', async () => {
```
Comment (npalm) [inline lambdas/functions/control-plane/src/aws/runners.test.ts:219]: I think this test case is not move, correct?

**Review by npalm — CHANGES_REQUESTED** [review]

File: lambdas/functions/control-plane/src/aws/runners.test.ts (hunk 11 above)
Comment (GuptaNavdeep1983) [reply lambdas/functions/control-plane/src/aws/runners.test.ts:219]: this is moved to scale-up.test.ts

File: lambdas/functions/control-plane/src/scale-runners/scale-up.ts [hunk 12]
```diff
@@ -139,8 +212,8 @@ export async function scaleUp(eventSource: string, payload: ActionRequestMessage
   if (eventSource !== 'aws:sqs') throw Error('Cannot handle non-SQS events!');
   const enableOrgLevel = yn(process.env.ENABLE_ORGANIZATION_RUNNERS, { default: true });
   const maximumRunners = parseInt(process.env.RUNNERS_MAXIMUM_COUNT || '3');
-  const runnerExtraLabels = process.env.RUNNER_EXTRA_LABELS;
-  const runnerGroup = process.env.RUNNER_GROUP_NAME;
+  const runnerLabels = process.env.RUNNER_LABELS || '';
```
[the remark this answers is not in the document]
Comment (GuptaNavdeep1983) [reply lambdas/functions/control-plane/src/scale-runners/scale-up.ts:215]: can be changed now.

File: lambdas/functions/control-plane/src/scale-runners/scale-up.ts (hunk 8 above)
Comment (GuptaNavdeep1983) [reply lambdas/functions/control-plane/src/scale-runners/scale-up.ts:149]: will create an issue.

Comment (GuptaNavdeep1983) [reply lambdas/functions/control-plane/src/scale-runners/scale-up.ts:149]: https://github.com/philips-labs/terraform-aws-github-runner/issues/3379

**Review by npalm — APPROVED** [review]

**Comment by mariusfilipowski** [comment]
Will this break support for GitHub Enterprise? There seems to be no support for JIT yet.

**Comment by jkruse14** [comment]
This breaks ephemeral runners. When running with `enable_ephemeral_runners = true`, I get:

```
<13>Jul 26 23:57:02 user-data: Starting runner after 0:1:32
<13>Jul 26 23:57:02 user-data: Starting the runner as user ec2-user
<13>Jul 26 23:57:02 user-data: Starting the runner in ephemeral mode
[   92.500730] cloud-init[2611]: ci-info: no authorized ssh keys fingerprints found for user ec2-user.
ci-info: no authorized ssh keys fingerprints found for user ec2-user.
<snip>
[   92.562938] cloud-init[2611]: Cloud-init v. 19.3-46.amzn2.0.1 finished at Wed, 26 Jul 2023 23:57:02 +0000. Datasource DataSourceEc2.  Up 92.55 seconds
<13>Jul 26 23:57:03 user-data: Unrecognized command-line input arguments for command run: 'ephemeral, url, token, runnergroup'. For usage refer to: .\config.cmd --help or ./config.sh --help
<13>Jul 26 23:57:03 user-data: An error occurred: Not configured. Run config.(sh/cmd) to configure the runner.
<13>Jul 26 23:57:03 user-data: Runner listener exit with terminated error, stop the service, no retry needed.
<13>Jul 26 23:57:03 user-data: Exiting runner...
```

Notice:
**user-data: Unrecognized command-line input arguments for command run: 'ephemeral, url, token, runnergroup'. For usage refer to: .\config.cmd --help or ./config.sh --help
**

setting `enable_ephemeral_runners = false` allows it to work. The difference is non-ephemeral runners do not use jitconfig

Also of note, I'm using the multi-runner config.

**Pull request merged**